functions/src/resource.ts

```diff
@@ -131,81 +149,107 @@ export class BaseResourceObject implements BaseResource {
-                  region
-                }
-                break;
-
-              case "iotOrganization":
-                if (!(tool === "dataFlow")) {
-                  reject(`Unknown iotOrganization tool: ${tool}`);
-                  return;
-                }
-                newResource = {
-                  type: "iotOrganization",
-                  tool,
-                  name,
-                  description,
-                  accessRules
-                }
-                break;
-
-              default:
-                reject(`Unknown resource type: ${type}`);
+      return BaseResourceObject.GetResourceSettings(db, env, type, tool)
+        .then((settings) => {
+          let newResource: FireStoreResource;
+          switch (type) {
+            case "s3Folder":
+              const {bucket, folder, region} = settings;
+              newResource = {
+                type: "s3Folder",
+                tool,
+                name,
+                description,
+                accessRules,
+                bucket,
+                folder,
+                region
+              };
+              break;
+
+            case "iotOrganization":
+              if (!(tool === "dataFlow")) {
+                reject(`Unknown iotOrganization tool: ${tool}`);
                 return;
-            }
+              }
+              newResource = {
+                type: "iotOrganization",
+                tool,
+                name,
+                description,
+                accessRules
+              };
+              break;
+
+            default:
+              reject(`Unknown resource type: ${type}`);
+              return;
+          }
 
-            return getResourceCollection(db, env).add(newResource)
-              .then((docRef) => docRef.get())
-              .then((docSnapshot) => BaseResourceObject.FromDocumentSnapshot(docSnapshot))
-              .then(resolve)
-              .catch(reject)
-          })
-          .catch(reject)
-      }
+          return getResourceCollection(db, env).add(newResource)
+            .then((docRef) => docRef.get())
+            .then((docSnapshot) => BaseResourceObject.FromDocumentSnapshot(docSnapshot))
+            .then(resolve)
+            .catch(reject)
+        })
+        .catch(reject)
     });
   }
 
-  static Update(db: FirebaseFirestore.Firestore, env: string, claims: JWTClaims, id: string, query: UpdateQuery) {
+  static Update(db: FirebaseFirestore.Firestore, env: string, claimsOrReadWriteToken: JWTClaims | string, id: string, query: UpdateQuery) {
     return new Promise<ResourceObject>((resolve, reject) => {
       const docRef = getResourceCollection(db, env).doc(id);
       return docRef.get()
         .then((docSnapshot) => {
           if (docSnapshot.exists) {
 
             const resource = BaseResourceObject.FromDocumentSnapshot(docSnapshot);
-            if (resource.isOwner(claims)) {
+            if (resource.canUpdate(claimsOrReadWriteToken)) {
```

Comment:
This is where it seems unsafe to allow a readWriteToken request to update the resource. On line 257 they could change the access rules and lock themselves out.